File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章最长公共子序列.py
#最长公共子序列  lanqiaoOJ 1189

n,m = map(int,input().split())
a = [0]+list(map(int,input().split()))
b = [0]+list(map(int,input().split()))
dp = [[0]*(m+1) for _ in range(2)]
now =0; old=1
for i in range(1,n+1):
    now,old=old,now
    for j in range(1,m+1):
        dp[now][j] = max(dp[now][j-1],dp[old][j])
        if a[i]==b[j]: dp[now][j]=max(dp[now][j],dp[old][j-1]+1)
print(dp[now][m])

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章最长公共子序列.cpp
//最长公共子序列   lanqiaoOJ 1189


#include<bits/stdc++.h>
using namespace std;
int dp[1001][1001],a[1001],b[1001];
int main(){
    int n,m;    cin>>n>>m;
    for(int i=1;i<=n;i++) scanf("%d",&a[i]);
    for(int i=1;i<=m;i++) scanf("%d",&b[i]);
    for(int i=1;i<=n;i++)
        for(int j=1;j<=m;j++) {
            if(a[i]==b[j]) dp[i][j] = max(dp[i][j],dp[i-1][j-1]+1);
            else           dp[i][j] = max(dp[i-1][j],dp[i][j-1]);

        }
   cout << dp[n][m];
}
